File: src/template/cpp/kmp/kmp.cpp
#include <vector>
#include <string>
using namespace std;

vector<int> prefix_function(const string& s) {
    int n = s.size();
    vector<int> pi(n);
    for (int i = 1; i < n; ++i) {
        int j = pi[i - 1];
        while (j > 0 && s[i] != s[j]) j = pi[j - 1];
        if (s[i] == s[j]) ++j;
        pi[i] = j;
    }
    return pi;
}

vector<int> kmp_search(const string& text, const string& pat) {
    string t = pat + "#" + text;
    auto pi = prefix_function(t);
    vector<int> res;
    int m = pat.size();
    for (int i = m + 1; i < (int)pi.size(); ++i) {
        if (pi[i] == m) res.push_back(i - 2 * m);
    }
    return res;
}

File: src/template/julia/kmp/kmp.jl
function prefix_function(s::AbstractString)
    n = lastindex(s)
    pi = zeros(Int, n)
    for i in 2:n
        j = pi[i-1]
        while j > 0 && s[i] != s[j+1]
            j = pi[j]
        end
        if s[i] == s[j+1]
            j += 1
        end
        pi[i] = j
    end
    return pi
end

function kmp_search(text::AbstractString, pat::AbstractString)
    t = pat * "#" * text
    pi = prefix_function(t)
    m = lastindex(pat)
    res = Int[]
    for i in m+2:length(pi)
        if pi[i] == m
            push!(res, i - 2m - 1)
        end
    end
    return res
end